File: backup/v2/ModbusAP.c
#include <stdio.h>
#include "ModbusAP.h"
#include "ModbusTCP.h"

int write_multiple_regs(const char *server_address, int port, uint16_t start_register, uint16_t num_registers, const uint16_t *values) {
    // Check consistency of parameters
    if (num_registers <= 0 || values == NULL) {
        return -1;
    }

    // Assemble APDU (MODBUS PDU)
    uint8_t apdu[256]; // Adjust the buffer size as needed
    // Fill in the APDU for a Modbus write operation (function code 0x10)
    apdu[0] = 0x10; // Function code for writing multiple registers
    apdu[1] = start_register >> 8; // High byte of starting register address
    apdu[2] = start_register & 0xFF; // Low byte of starting register address
    apdu[3] = num_registers >> 8; // High byte of number of registers to write
    apdu[4] = num_registers & 0xFF; // Low byte of number of registers to write
    apdu[5] = num_registers * 2; // Number of bytes to follow (2 bytes per register)

    // Copy the values to the APDU
    for (int i = 0; i < num_registers; i++) {
        apdu[6 + 2 * i] = values[i] >> 8; // High byte of register value
        apdu[7 + 2 * i] = values[i] & 0xFF; // Low byte of register value
    }

    // Send the Modbus request
    uint8_t apdu_response[256]; // Adjust the buffer size as needed
    int apdu_len = sizeof(apdu); // Calculate the actual APDU length
    int response_len = send_modbus_request(server_address, port, apdu, apdu_len, apdu_response);

    // Check the response (APDU_R or error_code)
    if (response_len < 0) {
        return -1;
    }

    // Print the response
    printf("Response: ");
    for (int i = 0; i < response_len; i++) {
        printf("%02X ", apdu_response[i]);
    }
    printf("\n");

    // Return the number of written registers or an error code
    return num_registers; // Replace with actual result
}

int read_h_regs(const char *server_address, int port, uint16_t start_register, uint16_t num_registers, uint16_t *values) {
    // Check consistency of parameters
    if (num_registers <= 0 || values == NULL) {
        return -1;
    }

    // Assemble APDU (MODBUS PDU)
    uint8_t apdu[256]; // Adjust the buffer size as needed
    // Fill in the APDU for a Modbus read operation (function code 0x03)
    apdu[0] = 0x03; // Function code for reading holding registers
    apdu[1] = start_register >> 8; // High byte of starting register address
    apdu[2] = start_register & 0xFF; // Low byte of starting register address
    apdu[3] = num_registers >> 8; // High byte of number of registers to read
    apdu[4] = num_registers & 0xFF; // Low byte of number of registers to read

    // Send the Modbus request
    uint8_t apdu_response[256]; // Adjust the buffer size as needed
    int apdu_len = sizeof(apdu); // Calculate the actual APDU length
    int response_len = send_modbus_request(server_address, port, apdu, apdu_len, apdu_response);

    // Check the response (APDU_R or error_code)
    if (response_len < 0) {
        return -1;
    }

    // Print the response
    printf("Response: ");
    for (int i = 0; i < response_len; i++) {
        printf("%02X ", apdu_response[i]);
    }
    printf("\n");

    // Return the number of read registers or an error code
    return num_registers; // Replace with actual result
}

File: ModbusAP.py
from ModbusTCP import send_modbus_request

def write_multiple_regs(server_address, port, start_register, num_registers, values):
    start_register = int(start_register, 16)
    num_registers = int(num_registers, 16)

    # Check consistency of parameters. # TODO: better check consistency of parameters. byten in message
    if num_registers <= 0 or not values:
        return -1

    # Assemble APDU.
    apdu = bytearray()
    apdu.append(0x10)
    apdu.extend([(start_register >> 8) & 0xFF, start_register & 0xFF])
    apdu.extend([(num_registers >> 8) & 0xFF, num_registers & 0xFF])
    apdu.append(num_registers * 2)

    for i in range(0, int(values[0:2]) * 2, 4):
        apdu.extend([(int(values[2+i:6+i], 16) >> 8) & 0xFF, int(values[2+i:6+i], 16) & 0xFF])

    apdu_len = len(apdu)

    # Send request.
    APDU_R, result = send_modbus_request(server_address, port, apdu, apdu_len)

    # Check response. TODO: better check response.

    # Return number of written registers
    if result == 0:
        num_written_registers = (APDU_R[-2] << 8) | APDU_R[-1]
        return num_written_registers        

    return -1

def read_holding_regs(server_address, port, start_register, num_registers):
    start_register = int(start_register, 16)
    num_registers = int(num_registers, 16)

    # Check consistency of parameters. # TODO: better check consistency of parameters. byten in message
    if num_registers <= 0:
        return -1

    # Assemble APDU
    apdu = bytearray()
    apdu.append(0x03)
    apdu.extend([(start_register >> 8) & 0xFF, start_register & 0xFF])
    apdu.extend([(num_registers >> 8) & 0xFF, num_registers & 0xFF])

    apdu_len = len(apdu)

    # Send request.
    APDU_R, result = send_modbus_request(server_address, port, apdu, apdu_len)

    values = bytearray()

    # Check response. TODO: better check response.

    # Return number of written registers
    if result == 0:
        num_read_registers = int(int(APDU_R[1]) / 2)
        
        for i in range(num_read_registers):
            values.append(APDU_R[2 + 2 * i])
            values.append(APDU_R[3 + 2 * i])
        
        return num_read_registers, values

    return -1

